### PR TITLE
Avoid pulling object function property augmentations when resolving intersections' properties

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/54345

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22351,8 +22351,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // from the target union, across all members
             const properties = getPropertiesOfType(target);
             const numericNamesOnly = isTupleType(source) && isTupleType(target);
-            for (const targetProp of excludeProperties(properties, excludedProperties)) {
+            for (let targetProp of excludeProperties(properties, excludedProperties)) {
                 const name = targetProp.escapedName;
+                if (target.flags & TypeFlags.Intersection) {
+                    targetProp = getPropertyOfUnionOrIntersectionType(target as IntersectionType, name, /*skipObjectFunctionPropertyAugment*/ true) || targetProp;
+                }
```
Comment (Andarist) [inline src/compiler/checker.ts:22358]: This is only needed for this extra properties check:
https://github.dev/microsoft/TypeScript/blob/f7e4f809acaea6b1b247ed05b2b7873d4546ab9e/src/compiler/checker.ts#L21338-L21349

But I'm also not sure if this is the right fix, nor if it's complete right now. I think it's important to establish what this should:
```ts
interface Test1 { toString: null | 'string'; }
type Test3 = Test1 & { optional?: unknown };

declare const test3: Test3
test3.toString // what this should be?
```

I'm not sure but perhaps this *shouldn't* be an intersection like `"string" & (() => string)`? Should this just be `null | 'string'`? If the latter then `skipObjectFunctionPropertyAugment` should be used on a different layer, when resolving the properties of an intersection type.

@RyanCavanaugh any thoughts about this? :)

**Review by sandersn — CHANGES_REQUESTED** [review]
You are right, skipObjectFunctionPropertyAugment should be used on a different layer.

**Comment by Andarist** [comment]
> You are right, skipObjectFunctionPropertyAugment should be used on a different layer.

Done.

File: tests/baselines/reference/mappedTypeAsClauses.types [hunk 2]
```diff
@@ -35,10 +35,10 @@ type TP1 = TypeFromDefs<{ name: 'a', type: string } | { name: 'b', type: number
...
...ay: string[]) => void, thisArg?: any) => void; getMap: () => <U>(callbackfn: (value: string, index: number, array: string[]) => U, thisArg?: any) => U[]; getFilter: () => { <S extends string>(predicate: (value: string, index: number, array: string[]) => value is S, thisArg?: any): S[]; (predicate: (value: string, index: number, array: string[]) => unknown, thisArg?: any): string[]; }; getReduce: () => { (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string): string; (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string, initialValue: string): string; <U>(callbackfn: (previousValue: U, currentValue: string, currentIndex: number, array: string[]) => U, initialValue: U): U; }; getReduceRight: () => { (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string): string; (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string, initialValue: string): string; <U>(callbackfn: (previousValue: U, currentValue: string, currentIndex: number, array: string[]) => U, initialValue: U): U; }; }
```
Comment (Andarist) [inline tests/baselines/reference/mappedTypeAsClauses.types:38]: both of the changes in this baseline are benign - properties were just reordered but nothing got added/lost

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13483,7 +13483,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             for (const current of type.types) {
                 for (const prop of getPropertiesOfType(current)) {
                     if (!members.has(prop.escapedName)) {
-                        const combinedProp = getPropertyOfUnionOrIntersectionType(type, prop.escapedName);
+                        const combinedProp = getPropertyOfUnionOrIntersectionType(type, prop.escapedName, /*skipObjectFunctionPropertyAugment*/ !!(type.flags & TypeFlags.Intersection));
```
Comment (sandersn) [inline src/compiler/checker.ts:13486]: why is this unconditional here but has a fallback in the `getPropertyOfType` code?

**Review by sandersn — CHANGES_REQUESTED** [review]
The results look right now but I don't understand why the two code changes are different.

File: src/compiler/checker.ts (hunk 3 above)
Comment (Andarist) [reply src/compiler/checker.ts:13486]: This resolves "own" members of the intersection so we never want to pull from function property augment here.

The other place is responsible for getting a `name` out of all accessible properties, it prioritizes its own ones but has to fallback to the ones from the augment to account for this:
```ts
interface Test1 { toString: null | 'string'; }
type Test2 = Test1 & { optional?: unknown };
declare const source: Test1;
const target: Test2 = { ...source };

const toString = target.toString; // own one, augment is avoided here
const hasOwn = target.hasOwnProperty; // not an own member but it should still be accessible
```

I pushed out this (`hasOwn` bit) as an extra thing in the added test case.

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by jakebailey** [comment]
Hm, doesn't feel unrelated when another user test suite reports the same error in a bunch of new repos...

**Comment by Andarist** [comment]
Hm, ye - definitely something is wrong here. Especially the DefinitelyTyped errors are interesting. I'll have to investigate this further later this week.

**Comment by sandersn** [comment]
material-ui check time is slower across node versions, by 1-2%. I'm not sure what the cutoff is for slowdowns.

**Comment by jakebailey** [comment]
The RWC script failed to post a comment, retrying with an updated token, but it does appear to have caused a break there.

@typescript-bot test this

**Comment by Andarist** [comment]
I pushed out a quick fix for at least one of the found issues. I plan to clean up this code later but it would be nice if you could re-run all of the extended test suites in the meantime.

**Comment by jakebailey** [comment]
RWC is still busted but I'll run everything and I can at least see if it failed. Really gotta get that infra replaced.

@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster
@typescript-bot pack this

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by Andarist** [comment]
There is still perf impact here that I have to investigate (and code cleanup) but it seems that now we are sitting only on the Ember-related breaks.

This break can be narrowed down to this:
```ts
declare class EmberObject {}

type PersonType = Readonly<typeof EmberObject> &
  (new (properties?: object) => {
    firstName: string;
    lastName: string;
  } & EmberObject) &
  (new (...args: any[]) => {
    firstName: string;
    lastName: string;
  } & EmberObject);

type PersonPrototype = PersonType["prototype"];
//   ^? 5.1: any; this PR: EmberObject
```

Since those assertions in Ember-related tests are doing things like:
```ts
assertType<string>(Person.prototype.firstName);
```
I'd call them broken in the first place because they just rely on `any`.

A mapped type (like `Readonly`) applied on a class loses its construct signatures but it doesn't lose properties like `prototype`. Let's make some tests:
```ts
type A = "prototype" extends keyof typeof EmberObject ? 1 : 0;
//  ^? type A = 1
type B = "prototype" extends keyof Readonly<typeof EmberObject> ? 1 : 0;
//  ^? type B = 1
type C = (typeof EmberObject)["prototype"];
//  ^? type C = EmberObject
type D = Readonly<typeof EmberObject>["prototype"];
//  ^? type D = EmberObject

type ConstructSignature = new (...args: any[]) => {
  firstName: string;
  lastName: string;
} & EmberObject;

type E = "prototype" extends ConstructSignature ? 1 : 0;
//  ^? type E = 0
type F = ConstructSignature["prototype"];
//  ^? type F = any  
```

So far we established that "own" properties (not from the object/function augment) should take priority over the augment in intersections. Given this desired semantic, this break looks desired to me. WDYT @sandersn ?

cc @chriskrycho @wagenet @gitKrystan @jamescdavis (Ember's typedef owners)

**Comment by wagenet** [comment]
I can verify that the existing tests are definitely broken. Types for EmberObject are significantly limited and the class will be going away entirely in the future so we were unlikely to actually fix those anyway.

**Comment by sandersn** [comment]
To check my understanding:

1. `assertType<string>(Person.prototype.firstName);` currently works because `Person.prototype.firstName: any`, right?
For the new code, I don't see a simple way to make an intersection propagate through `prototype` -- you'd have to add special-case code specifically for an object with `prototype` when intersected with an object with a `new` signature.

2. `type E = "prototype" extends ConstructSignature ? 1 : 0;` should be `extends keyof ConstructSignature`, right?

If the only breaks are in old versions of Ember, I'd guess that this pattern is not that common and is OK to not support (either correctly, or by inadvertent `any` like today).

**Comment by Andarist** [comment]
> assertType<string>(Person.prototype.firstName); currently works because Person.prototype.firstName: any, right?

Yes.

> For the new code, I don't see a simple way to make an intersection propagate through prototype -- you'd have to add special-case code specifically for an object with prototype when intersected with an object with a new signature.

Yes, that's probably what would have to be done. It would probably be good to break this down into some cases that might be desirable/practical. Do you have anything in particular in mind?

> type E = "prototype" extends ConstructSignature ? 1 : 0; should be extends keyof ConstructSignature, right?

Yes, my bad - it doesn't change the result though.

> If the only breaks are in old versions of Ember, I'd guess that this pattern is not that common and is OK to not support (either correctly, or by inadvertent any like today).

Should I assume then that this PR is something that you want to pull in and that I should investigate how to improve the perf?

**Comment by sandersn** [comment]
> Should I assume then that this PR is something that you want to pull in and that I should investigate how to improve the perf?

Yes. I don't think it's worth pursuing special case `prototype` in intersections.

**Pull request merged**